A80-pid.md

```diff
@@ -0,0 +1,290 @@
+  }
+
+  // use value calculated in the onEDFSchedulerUpdate method
+  meanUtilization = data.meanUtilization
+
+  // call the PID controller to get the value of the control signal.
+  controlSignal = data.pidController.update({
+    referenceSignal:  meanUtilization,
+    actualSignal:     utilization,
+    samplingInterval: time.Since(lastApplied),
+  })
+
+  // Normalize the signal.
+  // If meanUtilization ~= 0 the signal will be ~= 0 as well, and convergence will become painfully slow.
+  // If, meanUtilization >> 1 the signal may become very high, which could lead to oscillations.
+  if meanUtilization > 0 {
+    controlSignal *= 1 / meanUtilization
+  }
+
+  lastAppliedWeight = data.lastAppliedWeightPerSubchannel[subchannelID]
+
+  // Use controlSignal to adjust the weight.
+  // First calculate a multiplier that will be used to determine how much weight should be changed.
+  // The higher is the absolute value of the controlSignal the more we need to adjust the weight.
+  if controlSignal >= 0 {
+    // in this case mult should belong to the [1,inf) interval, so we will be increasing the weight.
+    mult = 1.0 + controlSignal
+  } else {
+    // in this case mult should belong to (0, 1) interval, so we will be decreasing the weight.
+    mult = -1.0 / (controlSignal - 1.0)
+  }
+  weight = lastAppliedWeight * mult
+
+  // Clamp weight
+  if weight > conf.MaxWeight {
+    weight = conf.MaxWeight
+  }
+  if weight < conf.MinWeight {
+    weight = conf.MinWeight
+  }
+
+  // Save resulting utilization and weight.
+  data.utilizationPerSubchannel[subchannelId] = utilization
+  data.lastAppliedWeightPerSubchannel[subchannelID] = weight
+
+  return weight
+}
+
+func onEDFSchedulerUpdate(data callbackData) {
+  // Calculate mean utilization across all subchannels
+  totalUtilization = 0
+  count = len(data.utilizationPerSubchannel)
+  for _, utilization in data.utilizationPerSubchannel {
+    totalUtilization += utilization
+  }
+  data.meanUtilization = totalUtilization / count
+}
+```
+
+The proposal is to make `wrrCallbacks` public. This has a number of significant benefits. Besides PID, there are other cases where one might need to extend `wrr`. For example, Spotify [demonstrates](https://www.youtube.com/watch?v=8E5zVdEfwi0) a gRPC load balancer to reduce cross-zone traffic – this can be implemented nicely in terms of `wrr` weights. We are also considering the same and incorporating things like latency into our load balancing decisions. Existing ORCA extension points don't cover these use cases. We leverage ORCA for custom server utilization metrics, but we also need the ability to combine server and client metrics to generate the resulting weight. The alternative is to write our own balancer with custom EDF scheduler and handle details related to subchannel management and interactions with resolvers. With this new API, use cases like this can be covered naturally, users have full control over the end-to-end definition of weights.
```

Comment:
cc @yannickepstein would something like the proposed wrrCallbacks interface  work for your use-case, so you don't have to copy-paste the whole wrr?